google/cloud/storage/internal/http_response.cc

```diff
@@ -50,87 +50,147 @@ Status AsStatus(HttpResponse const& http_response) {
+    return StatusCode::kOutOfRange;
   }
-  if (http_response.status_code ==
-      HttpStatusCode::kRequestRangeNotSatisfiable) {
-    return Status(StatusCode::kOutOfRange, http_response.payload);
+  if (code == HttpStatusCode::kRequestRangeNotSatisfiable) {
+    return StatusCode::kOutOfRange;
   }
-  if (http_response.status_code == HttpStatusCode::kTooManyRequests) {
-    return Status(StatusCode::kUnavailable, http_response.payload);
+  if (code == HttpStatusCode::kTooManyRequests) {
+    return StatusCode::kUnavailable;
   }
-  if (HttpStatusCode::kMinRequestErrors <= http_response.status_code &&
-      http_response.status_code < HttpStatusCode::kMinInternalErrors) {
+  if (HttpStatusCode::kMinRequestErrors <= code &&
+      code < HttpStatusCode::kMinInternalErrors) {
     // 4XX - A request error.
-    return Status(StatusCode::kInvalidArgument, http_response.payload);
+    return StatusCode::kInvalidArgument;
   }
-  if (http_response.status_code == HttpStatusCode::kInternalServerError) {
-    return Status(StatusCode::kUnavailable, http_response.payload);
+  if (code == HttpStatusCode::kInternalServerError) {
+    return StatusCode::kUnavailable;
   }
-  if (http_response.status_code == HttpStatusCode::kBadGateway) {
-    return Status(StatusCode::kUnavailable, http_response.payload);
+  if (code == HttpStatusCode::kBadGateway) {
+    return StatusCode::kUnavailable;
   }
-  if (http_response.status_code == HttpStatusCode::kServiceUnavailable) {
-    return Status(StatusCode::kUnavailable, http_response.payload);
+  if (code == HttpStatusCode::kServiceUnavailable) {
+    return StatusCode::kUnavailable;
   }
-  if (HttpStatusCode::kMinInternalErrors <= http_response.status_code &&
-      http_response.status_code < HttpStatusCode::kMinInvalidCode) {
+  if (HttpStatusCode::kMinInternalErrors <= code &&
+      code < HttpStatusCode::kMinInvalidCode) {
     // 5XX - server errors are mapped to kInternal.
-    return Status(StatusCode::kInternal, http_response.payload);
+    return StatusCode::kInternal;
   }
-  return Status(StatusCode::kUnknown, http_response.payload);
+  return StatusCode::kUnknown;
+}
+
+// Makes an `ErrorInfo` from an `"error"` JSON object that looks like
+//   [
+//     {
+//       "@type": "type.googleapis.com/google.rpc.ErrorInfo",
+//       "reason": "..."
+//       "domain": "..."
+//       "metdata": {
+//         "key1": "value1"
+//         ...
+//       }
+//     }
+//   ]
+// See also https://cloud.google.com/apis/design/errors#http_mapping
+ErrorInfo MakeErrorInfo(nlohmann::json const& details) {
+  static auto const* const kErrorInfoType =
+      "type.googleapis.com/google.rpc.ErrorInfo";
+  for (auto const& e : details.items()) {
+    auto const& v = e.value();
+    if (v.value("@type", "") != kErrorInfoType) continue;
+    auto reason = v.value("reason", "");
+    auto domain = v.value("domain", "");
+    auto metadata_json = v.value("metadata", nlohmann::json::object());
```

Comment:
nit: I do not worry about copying here, this should be the uncommon path.  But if one was worried about copying, there is a copy of something like a small `std::map<>`.  You could do something like:

```cc
  auto domain = v.value("domain", "");
  if (!v.contains("metadata")) return ErrorInfo{std::move(reason), std::move(domain), {});
  auto metadata = std::unordered_map<std::string, std::string>{};
  for (auto const& m : v["metadata"].items()) {
    metadata[m.key()] = m.value();
  }
...
```

Reply:
Thanks. Due to the likely minimal copying that would save, I think I'll leave the code as is, because I think it's possibly slightly easier to read as is.